pkg/auth/providers/azure/clients/ad_graph_client.go

```diff
@@ -81,55 +81,53 @@ func (c azureADGraphClient) ListGroupMemberships(id string) ([]string, error) {
 	return *groups.Value, nil
 }
 
-func (c azureADGraphClient) LoginUser(credential *v32.AzureADLogin, config *v32.AzureADConfig, _ corev1.SecretInterface) (v3.Principal, []v3.Principal, string, error) {
+func (c *azureADGraphClient) LoginUser(credential *v32.AzureADLogin, config *v32.AzureADConfig, _ corev1.SecretInterface) (v3.Principal, []v3.Principal, string, error) {
 	logrus.Debug("[AZURE_PROVIDER] Started token swap with AzureAD")
 
-	azureClient, err := NewClientFromCode(credential.Code, config)
+	err := c.setInternalFields(credential.Code, config)
 	if err != nil {
 		return v3.Principal{}, nil, "", err
 	}
 
 	logrus.Debug("[AZURE_PROVIDER] Completed token swap with AzureAD")
 
-	oid, err := ExtractFieldFromJWT(azureClient.AccessToken(), "oid")
+	oid, err := ExtractFieldFromJWT(c.AccessToken(), "oid")
 	if err != nil {
 		return v3.Principal{}, nil, "", err
 	}
 
 	logrus.Debug("[AZURE_PROVIDER] Started getting user info from AzureAD")
 
-	userPrincipal, err := azureClient.GetUser(oid.(string))
+	userPrincipal, err := c.GetUser(oid.(string))
 	if err != nil {
 		return v3.Principal{}, nil, "", err
 	}
 	userPrincipal.Me = true
 	logrus.Debug("[AZURE_PROVIDER] Completed getting user info from AzureAD")
 
-	userGroups, err := azureClient.ListGroupMemberships(GetPrincipalID(userPrincipal))
+	userGroups, err := c.ListGroupMemberships(GetPrincipalID(userPrincipal))
 	if err != nil {
 		return v3.Principal{}, nil, "", err
 	}
 
-	groupPrincipals, err := UserGroupsToPrincipals(azureClient, userGroups)
+	groupPrincipals, err := UserGroupsToPrincipals(c, userGroups)
 	if err != nil {
 		return v3.Principal{}, nil, "", err
 	}
 
-	providerToken, err := azureClient.MarshalTokenJSON()
+	providerToken, err := c.MarshalTokenJSON()
 	if err != nil {
 		return v3.Principal{}, nil, "", err
 	}
 
 	return userPrincipal, groupPrincipals, providerToken, nil
 }
 
-// NewClientFromCode sets up the SPT, user and group client using a code
-func NewClientFromCode(code string, config *v32.AzureADConfig) (AzureClient, error) {
-	ac := &azureADGraphClient{}
-
+// setInternalFields configures the SPT, user, and group client using a code.
+func (c *azureADGraphClient) setInternalFields(code string, config *v32.AzureADConfig) error {
```

Comment:
Near as I can tell, you really have two separate constructors here:
- One constructor takes in an auth code and returns a constructed client
- One constructor takes in a token and returns a constructed client

I would recommend creating this as a separate constructor and then using that constructor when the provider's login method is called. There's also a fair amount of duplicated logic between these two constructors - I would consider moving that common logic into a common function.